File: android/app/src/main/java/com/mccontroller/core/ControllerMode.kt
package com.mccontroller.core

/**
 * Mode pushed by the PC server based on detected MC window/cursor state.
 * Drives which Android UI layer is visible.
 */
enum class ControllerMode(val byteValue: Byte) {
    /** MC focused + GLFW cursor captured. Full controller (joystick + buttons). */
    InGame(0),

    /** MC focused but cursor visible. LookPad drives cursor; reduced button set. */
    UiInteract(1),

    /** MC not in foreground. Lock screen overlay; reject most input. */
    AntiMistouch(2);

    companion object {
        fun fromByte(b: Byte): ControllerMode = when (b.toInt()) {
            0 -> InGame
            1 -> UiInteract
            else -> AntiMistouch  // 2 or unknown → safe default
        }
    }
}

File: ios/McController/Core/ControllerMode.swift
import Foundation

/// 3-state server-driven mode.
///
/// The PC server polls Minecraft's foreground state + cursor visibility every
/// 100ms and pushes a STATE_CHANGE whenever the derived mode changes. The
/// client renders different controllers per mode and resets all gesture FSMs
/// on transition.
enum ControllerMode: UInt8, Sendable, Equatable {
    case inGame        = 0  // full controller (joystick + buttons + hotbar)
    case uiInteract    = 1  // LookPad drives cursor, reduced button set
    case antiMistouch  = 2  // lock overlay; input blocked

    init(wireByte: UInt8) {
        self = ControllerMode(rawValue: wireByte) ?? .antiMistouch
    }
}
